src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ShardedReadsDatasource.java

```diff
@@ -0,0 +1,361 @@
+                                        outputting.stop();
+                                        bunny.stepEnd("output subshard " + currentShardIndex + ", " + ret.size() + " reads");
+                                        shipping.stop();
+                                        choosing.start();
+                                        ret.clear();
+                                    }
+                                    currentShardIndex++;
+                                    currentShard = shards.get(currentShardIndex);
+                                }
+
+                                // work around some serialization madness
+                                r.setHeader(null);
+                                ret.add(new SAMRecordToGATKReadAdapter(r));
+                            }
+                            buf.clear();
+                            batcher.reuse(buf);
+                            choosing.stop();
+                            loading.start();
+                        }
+                        bunny.stepEnd("that was the last batch");
+                        loading.stop();
+                    } else {
+                        // TODO: HDFS input
+                        throw new RuntimeException("Sorry, Hadoop support not yet implemented");
+                    }
+                    shipping.start();
+                    if (!ret.isEmpty()) {
+                        kving.start();
+                        accepted += ret.size();
+                        KV<String, Iterable<GATKRead>> out = KV.of(currentShard.toString(), ret);
+                        kving.stop();
+                        outputting.start();
+                        c.output(out);
+                        outputting.stop();
+                        //bunny.stepEnd("output " + ret.size() + " reads");
+                    }
+                    shipping.stop();
+                    readsEmitted += accepted;
+                    //bunny.stepEnd("loading " + accepted + " reads for " + i.toString()
+                    //    + " in " + processing.elapsed(TimeUnit.MILLISECONDS) + " ms. "
+                    //    + " readingInput: " + readingInput.elapsed(TimeUnit.MILLISECONDS) + ", "
+                    //    + " init: " + init.elapsed(TimeUnit.MILLISECONDS) + ", "
+                    //    + " cutShards: " + cutShards.elapsed(TimeUnit.MILLISECONDS) + ", "
+                    //    + " kv.of: " + kving.elapsed(TimeUnit.MILLISECONDS) + ", "
+                    //    + " output: " + outputting.elapsed(TimeUnit.MILLISECONDS) + ", "
+                    //    + " rejected " + rejected);
+                }
+
+                private boolean accept(GATKRead r, SimpleInterval region) {
+                    if (r.isUnmapped()) {
+                        // we don't want unmapped reads, *unless* their mate is in the region
+                        return (!r.mateIsUnmapped() && r.getMateStart() >= region.getStart());
```

Comment:
Rather than hardcoding this filtering in the data source like this, you should allow the datasource to take a set of standard hellbender `ReadFilters`, and make the default filter(s) do what you want.